go/vt/wrangler/traffic_switcher.go

```diff
@@ -1910,6 +1942,280 @@ func (ts *trafficSwitcher) isSequenceParticipating(ctx context.Context) (bool, e
 	return sequenceFound, nil
 }
 
+// getTargetSequenceMetadata returns a map of sequence metadata keyed by the
+// backing sequence table name. If the target keyspace has no tables
+// defined that use sequences for auto_increment generation then a nil
+// map will be returned.
+func (ts *trafficSwitcher) getTargetSequenceMetadata(ctx context.Context) (map[string]*sequenceMetadata, error) {
+	vschema, err := ts.TopoServer().GetVSchema(ctx, ts.targetKeyspace)
+	if err != nil {
+		return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "failed to get vschema for target keyspace %s: %v",
+			ts.targetKeyspace, err)
+	}
+	if vschema == nil || len(vschema.Tables) == 0 { // Nothing to do
+		return nil, nil
+	}
+
+	sequencesByBackingTable, backingTablesFound, err := ts.findSequenceUsageInKeyspace(vschema)
+	if err != nil {
+		return nil, err
+	}
+	// If all of the sequence tables were defined using qualified table
+	// names then we don't need to search for them in other keyspaces.
+	if len(sequencesByBackingTable) == 0 || backingTablesFound {
+		return sequencesByBackingTable, nil
+	}
+
+	if err := ctx.Err(); err != nil {
+		return nil, err
+	}
+
+	// Now we need to locate the backing sequence table(s) which will
+	// be in another unsharded keyspace.
+	smMu := sync.Mutex{}
+	tableCount := len(sequencesByBackingTable)
+	tablesFound := 0                                                      // Used to short circuit the search
+	searchCompleted := make(chan struct{})                                // The search has completed
+	searchKeyspace := func(sctx context.Context, keyspace string) error { // The function used to search each keyspace
+		kvs, kerr := ts.TopoServer().GetVSchema(sctx, keyspace)
+		if kerr != nil {
+			return vterrors.Errorf(vtrpcpb.Code_INTERNAL, "failed to get vschema for keyspace %s: %v",
+				keyspace, kerr)
+		}
+		if kvs == nil || kvs.Sharded || len(kvs.Tables) == 0 {
+			return nil
+		}
+		for tableName, tableDef := range kvs.Tables {
+			select {
+			case <-sctx.Done():
+				return sctx.Err()
+			case <-searchCompleted:
+				return nil
+			default:
+			}
+			if complete := func() bool {
```

Comment:
This function is called only once within each call to `searchKeyspace`. It seems overkill to define it as an inner function here, the code could as well have been written out procedurally with better clarity.
The other option is to define it as a function outside `searchKeyspace` and simply call it in the scope of `searchKeyspace`. That would have the merit of making `searchKeyspace` shorter and easier to read.
Also, using local variables within inner functions can be a cause of future errors. Bearing that in mind, extracting it out and enumerating all the function parameters and passing them explicitly will be good. But it's also fine to just get rid of the indirection of a function definition.

Reply:
I now see the prior comments that motivated this pattern. I still think the closure should be extracted out so that the local variable entanglement can be avoided.